src/graphics/vulkan.zig

```diff
@@ -345,3 +351,139 @@ fn createLogicalDevice() void {
 	c.vkGetDeviceQueue(device, indices.graphicsFamily.?, 0, &graphicsQueue);
 	c.vkGetDeviceQueue(device, indices.presentFamily.?, 0, &presentQueue);
 }
+
+const SwapChain = struct { // MARK: SwapChain
+	var swapChain: c.VkSwapchainKHR = null;
+	var images: []c.VkImage = undefined;
+	var imageViews: []c.VkImageView = undefined;
+	var imageFormat: c.VkFormat = undefined;
+	var extent: c.VkExtent2D = undefined;
+
+	const SupportDetails = struct {
+		capabilities: c.VkSurfaceCapabilitiesKHR,
+		formats: []const c.VkSurfaceFormatKHR,
+		presentModes: []const c.VkPresentModeKHR,
+
+		fn init(allocator: NeverFailingAllocator, physical: c.VkPhysicalDevice) SupportDetails {
+			var result: SupportDetails = undefined;
+			checkResult(c.vkGetPhysicalDeviceSurfaceCapabilitiesKHR(physical, surface, &result.capabilities));
+			result.formats = getPhysicalDeviceSurfaceFormatsKHR(allocator, physical);
+			result.presentModes = getPhysicalDeviceSurfacePresentModesKHR(allocator, physical);
+			return result;
+		}
+
+		fn deinit(self: SupportDetails, allocator: NeverFailingAllocator) void {
+			allocator.free(self.formats);
+			allocator.free(self.presentModes);
+		}
+
+		fn chooseFormat(self: SupportDetails) c.VkSurfaceFormatKHR {
+			for(self.formats) |format| {
+				if(format.format == c.VK_FORMAT_B8G8R8A8_SRGB and format.colorSpace == c.VK_COLOR_SPACE_SRGB_NONLINEAR_KHR) {
+					return format;
+				}
+			}
+			std.log.err("Couldn't find swapchain format BGRA8, choosing any instead", .{});
+			return self.formats[0];
+		}
+
+		fn chooseSwapPresentMode(self: SupportDetails) c.VkPresentModeKHR {
+			_ = self; // TODO: Use MAILBOX if vsync is disabled
+			return c.VK_PRESENT_MODE_FIFO_KHR;
+		}
+
+		fn chooseSwapExtent(self: SupportDetails) c.VkExtent2D {
+			if(self.capabilities.currentExtent.width != std.math.maxInt(u32)) {
+				return self.capabilities.currentExtent;
+			}
+			var width: i32 = undefined;
+			var height: i32 = undefined;
+			c.glfwGetFramebufferSize(main.Window.vulkanWindow, &width, &height);
+			return .{
+				.width = @min(self.capabilities.maxImageExtent.width, @max(self.capabilities.minImageExtent.width, @max(0, width))),
+				.height = @min(self.capabilities.maxImageExtent.height, @max(self.capabilities.minImageExtent.height, @max(0, height))),
+			};
+		}
+	};
+
+	fn createImageView(image: c.VkImage) c.VkImageView {
```

Comment:
I suspect that creating image views will be a common operation. We may want to get ourselves an VkImage abstraction (Image struct) object with view() method to do create views. I'm not saying that this is right moment to do so, but it may be useful to keep that in mind for the future.